fs/copy.py

```diff
@@ -126,105 +147,135 @@ def copy_file(
         dst_path (str): Path to a file on the destination filesystem.
+        condition (str): Name of the condition to check for each file.
 
     Returns:
         bool: `True` if the file copy was executed, `False` otherwise.
 
     """
     with manage_fs(src_fs, writeable=False) as _src_fs:
         with manage_fs(dst_fs, create=True) as _dst_fs:
-            if _src_fs is _dst_fs:
-                # Same filesystem, so we can do a potentially optimized
-                # copy
-                if _source_is_newer(_src_fs, src_path, _dst_fs, dst_path):
-                    _src_fs.copy(src_path, dst_path, overwrite=True)
-                    return True
-                else:
-                    return False
-            else:
-                # Standard copy
-                with _src_fs.lock(), _dst_fs.lock():
-                    if _source_is_newer(_src_fs, src_path, _dst_fs, dst_path):
-                        copy_file_internal(_src_fs, src_path, _dst_fs, dst_path)
-                        return True
-                    else:
-                        return False
+            do_copy = _copy_is_necessary(
+                _src_fs, src_path, _dst_fs, dst_path, condition
+            )
+            if do_copy:
+                copy_file_internal(_src_fs, src_path, _dst_fs, dst_path, True)
+            return do_copy
+
+
+def copy_file_internal(
+    src_fs,  # type: FS
+    src_path,  # type: Text
+    dst_fs,  # type: FS
+    dst_path,  # type: Text
+    lock=False,  # type: bool
+):
+    # type: (...) -> None
+    """Copy a file at low level, without calling `manage_fs` or locking.
+
+    If the destination exists, and is a file, it will be first truncated.
+
+    This method exists to optimize copying in loops. In general you
+    should prefer `copy_file`.
+
+    Arguments:
+        src_fs (FS): Source filesystem.
+        src_path (str): Path to a file on the source filesystem.
+        dst_fs (FS): Destination filesystem.
+        dst_path (str): Path to a file on the destination filesystem.
+        lock (bool): Lock both filesystems before copying.
+
+    """
+    if src_fs is dst_fs:
+        # Same filesystem, so we can do a potentially optimized
+        # copy
+        src_fs.copy(src_path, dst_path, overwrite=True)
+        return
+
+    def _copy_locked():
+        if dst_fs.hassyspath(dst_path):
+            with dst_fs.openbin(dst_path, "w") as write_file:
+                src_fs.download(src_path, write_file)
+        else:
+            with src_fs.openbin(src_path) as read_file:
+                dst_fs.upload(dst_path, read_file)
+
+    if lock:
+        with src_fs.lock(), dst_fs.lock():
+            _copy_locked()
+    else:
+        _copy_locked()
 
 
 def copy_structure(
     src_fs,  # type: Union[FS, Text]
     dst_fs,  # type: Union[FS, Text]
     walker=None,  # type: Optional[Walker]
+    src_root="/",  # type: Text
+    dst_root="/",  # type: Text
```

Comment:
A question here: do we really need these new `src_root` and `dst_root` parameters? It's already possible to copy subdirectories using the `opendir` and `makedir` to get a `SubFS`:
```python
src_fs = fs.open_fs("...")
dst_fs = fs.open_fs("...")
fs.copy_structure(fs.opendir("foo"), fs.makedir("bar", recreate=False))
```
I don't know if we really need to add 2 extra parameters, it may get confusing if there is more than one way to do the same thing (aka *There should be one-- and preferably only one --obvious way to do it.*).

Reply:
Sure, if you feel like those parameters should be covered by `SubFS` instead, I can do that. Although the question is whether the break with the other functions in the module isn't a more confusing design. (e.g. `copy_dir` takes both `src_fs` and `src_path`).

Reply:
True. Then in that case, maybe having the signature be `copy_structure(src_fs, src_path, dst_fs, dst_path, walker=None)` would be more consistent?

Reply:
I agree but that would be a breaking change. Would that be acceptable?

Reply:
@althonos Any opinion on this? I could easily change the signature of that function but it feels to me like that would break too much existing code.

Reply:
Let's stick with the non-breaking signature for now, it looks like we are going to have to refactor `fs.copy` in v3 anyway.